common/djangoapps/third_party_auth/pipeline.py

```diff
@@ -739,16 +740,29 @@ def associate_by_email_if_login_api(auth_entry, backend, details, user, current_
     if auth_entry == AUTH_ENTRY_LOGIN_API:
         # Temporary custom attribute to help ensure there is no usage.
         set_custom_attribute('deprecated_auth_entry_login_api', True)
-        association_response = associate_by_email(backend, details, user, *args, **kwargs)
-        if (
-            association_response and
-            association_response.get('user') and
-            association_response['user'].is_active
-        ):
-            # Only return the user matched by email if their email has been activated.
-            # Otherwise, an illegitimate user can create an account with another user's
-            # email address and the legitimate user would now login to the illegitimate
-            # account.
+
+        association_response, user_is_active = get_associated_user_by_email_response(
+            backend, details, user, *args, **kwargs)
+
+        if user_is_active:
+            return association_response
+
+
+@partial.partial
+def associate_by_email_if_oauth(auth_entry, backend, details, user, strategy, *args, **kwargs):
+    """
+    This pipeline step associates the current social auth with the user with the
+    same email address in the database.  It defers to the social library's associate_by_email
+    implementation, which verifies that only a single database user is associated with the email.
+
+    This association is done ONLY if the user entered the pipeline belongs to Oauth provider.
+    """
+
+    if is_oauth_provider(backend.name, **kwargs):
```

Comment:
It will enable it by default for each auth entry in the oauth2 provider's case and I believe the purpose of `associate_by_email_if_login_api` is failed here. Don’t you think we need to put it under the feature flag `ENABLE_REQUIRE_THIRD_PARTY_AUTH`?

Reply:
And if we are making it the default behavior in the oauth2 providers case, then I think we can change this check `if auth_entry == AUTH_ENTRY_LOGIN_API:` to something `if is_oauth_provider(backend.name, **kwargs) and auth_entry in ALLOWED_AUTH_ENTRIES:` instead of this new pipeline function.

`ALLOWED_AUTH_ENTRIES` will be a list of auth entries you want to allow here. e.g. `login_api`, `login`, `register`.

Reply:
@waheedahmed Yes, I do think we need to put it under the feature flag `ENABLE_REQUIRE_THIRD_PARTY_AUTH`, that's a great point.

Your recommendation about the `ALLOWED_AUTH_ENTRIES` is great, but I'm hesitant to make such a change because it can be considered "breaking".

Reply:
@waheedahmed may you please check the latest changes? I added the setting check. I also added a test to verify that it is working 😃 

I thought of adding another test, when the flag is disabled, but I believe `test_new_account_registration_assigns_distinct_username_on_collision` already covers the default case. Let me know though, please, if you don't consider that to be enough 👍🏼